packages/@aws-cdk/core/lib/annotations.ts

```diff
@@ -49,6 +51,38 @@ export class Annotations {
     this.addMessage(cxschema.ArtifactMetadataEntryType.ERROR, message);
   }
 
+  /**
+   * Adds a deprecation warning for a specific API.
+   *
+   * Deprecations will be added only once per construct as a warning and will be
+   * deduplicated based on the `api`.
+   *
+   * If the environment variable `CDK_BLOCK_DEPRECATIONS` is set, this method
+   * will throw an error instead with the deprecation message.
+   *
+   * @param api The API being deprecated in the format `module.Class.property`
+   * (e.g. `@aws-cdk/core.Construct.node`).
+   * @param message The deprecation message to display, with information about
+   * alternatives.
+   */
+  public addDeprecation(api: string, message: string) {
+    const text = `The API ${api} is deprecated: ${message}. This API should will be removed in the next major version of the AWS CDK`;
+
+    // throw if CDK_BLOCK_DEPRECATIONS is set
+    if (process.env.CDK_BLOCK_DEPRECATIONS) {
+      throw new Error(`${this.scope.construct.path}: ${text}`);
+    }
```

Comment:
Can we change the behaviour such that all deprecations are reported when `CDK_BLOCK_DEPRECATIONS` is set, rather than one by one?

It seems like we'll benefit from a `getDeprecations()` API rather than that coalesce this with all warnings.

Reply:
Feel like over-design. The main goal of `CDK_BLOCK_DEPRECATIONS` is to protect _us_ against usage of deprecated APIs in the CDK code. It's not a user facing feature at this point.

Reply:
Lets see how this plays out and introduce this later if needed. Ok?

Reply:
Why the difference between _us_ vs _customers_? Why not build one mechanism for both?

We could use this to report use of deprecated APIs on the CLI or error out (depending on customer's choice) as a preparation for our customers to migrate to v2.

Reply:
Customers will see deprecation warnings in the CLI and can use the `--strict` option to fail on warning (if I recall).

So users are covered.

The environment variable is a "fail fast" approach for our tests